OpenRA.Game/Support/ConditionExpression.cs

```diff
@@ -0,0 +1,901 @@
+
+		class AstStack
+		{
+			readonly List<Expression> expressions = new List<Expression>();
+			readonly List<ExpressionType> types = new List<ExpressionType>();
+
+			public ExpressionType PeekType() { return types[types.Count - 1]; }
+
+			public Expression Peek(ExpressionType toType)
+			{
+				var fromType = types[types.Count - 1];
+				var expression = expressions[expressions.Count - 1];
+				if (toType == fromType)
+					return expression;
+
+				switch (fromType)
+				{
+					case ExpressionType.Bool:
+						return IfThenElse(expression, One, Zero);
+					case ExpressionType.Int:
+						return IfThenElse(AsBool(expression), True, False);
+				}
+
+				throw new InvalidProgramException("Unable to convert ExpressionType.{0} to ExpressionType.{1}".F(
+					Enum<ExpressionType>.GetValues()[(int)fromType], Enum<ExpressionType>.GetValues()[(int)toType]));
+			}
+
+			public Expression Pop(ExpressionType type)
+			{
+				var expression = Peek(type);
+				expressions.RemoveAt(expressions.Count - 1);
+				types.RemoveAt(types.Count - 1);
+				return expression;
+			}
+
+			public void Push(Expression expression, ExpressionType type)
+			{
+				expressions.Add(expression);
+				if (type == ExpressionType.Int)
+					if (expression.Type != typeof(int))
+						throw new InvalidOperationException("Expected System.Int type instead of {0} for {1}".F(expression.Type, expression));
+
+				if (type == ExpressionType.Bool)
+					if (expression.Type != typeof(bool))
+						throw new InvalidOperationException("Expected System.Boolean type instead of {0} for {1}".F(expression.Type, expression));
+				types.Add(type);
+			}
+
+			public void Push(Expression expression)
+			{
+				expressions.Add(expression);
+				if (expression.Type == typeof(int))
+					types.Add(ExpressionType.Int);
+				else if (expression.Type == typeof(bool))
+					types.Add(ExpressionType.Bool);
+				else
+					throw new InvalidOperationException("Unhandled result type {0} for {1}".F(expression.Type, expression));
+			}
+
+			public bool BinaryOperandsOnTopAreBoolean()
+			{
+				var y = types[types.Count - 1];
+				var x = types[types.Count - 2];
+				return y == ExpressionType.Bool && x == ExpressionType.Bool;
+			}
+		}
+
+		class Compiler
+		{
+			readonly AstStack ast = new AstStack();
+
+			void BooleanJunction(TokenType type)
+			{
+				if (ast.BinaryOperandsOnTopAreBoolean())
+				{
+					var y = ast.Pop(ExpressionType.Bool);
+					var x = ast.Pop(ExpressionType.Bool);
+					if (type == TokenType.And)
+						ast.Push(Expressions.Expression.And(x, y));
+					else if (type == TokenType.Or)
+						ast.Push(Expressions.Expression.Or(x, y));
+					else
+						throw new InvalidProgramException("void BooleanJunction(TokenType type) : type must be And or Or");
+				}
+				else
+				{
+					var y = ast.Pop(ExpressionType.Int);
+					var test = ast.Peek(ExpressionType.Bool);
+					var x = ast.Pop(ExpressionType.Int);
+					if (type == TokenType.And)
+						ast.Push(IfThenElse(test, y, x));
```

Comment:
I don't think python's `and`/`or` behaviour is a good fit here.  It is useful in python as a bodge for a null/falsy-coalescing operator on arbitrary objects, but I find it to be counterintuitive for ints.

Can we please return 1 or 0 instead, for consistency with C/C++?